src/main/java/patternverifiers/SingletonVerifier.java

```diff
@@ -1,17 +1,312 @@
 package patternverifiers;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
 import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.Modifier;
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.body.ConstructorDeclaration;
+import com.github.javaparser.ast.body.FieldDeclaration;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.expr.BinaryExpr;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.ast.expr.ObjectCreationExpr;
+import com.github.javaparser.ast.stmt.IfStmt;
+
+import base.VariableReader;
 
 /**
  * A verifier for the singleton pattern.
  */
 public class SingletonVerifier implements IPatternVerifier {
 
+    private final transient List<ConstructorDeclaration> constDeclList = new ArrayList<>();
+    private transient FieldDeclaration instanceVar;
+
     public SingletonVerifier() {
     }
 
     @Override
     public Feedback verify(CompilationUnit compUnit) {
-        throw new UnsupportedOperationException();
+        boolean isValid =
+            callsConstructor(compUnit) && hasAStaticInstance(compUnit) && onlyInstantiatedIfNull(
+            compUnit) && hasPrivateConstructor(compUnit);
+
+        return new Feedback(isValid);
```

Comment:
Would be nice if a message explaining which predicate failed and why.

Reply:
Should the part of the verify method in ImmutableVerifier be moved to a separate class to make allow reuse?

Reply:
Reminder to update to ClassOrInterfaceDeclaration from CompilationUnit